src/purple_vm/src/primitives/value.rs

```diff
@@ -108,141 +201,1118 @@ impl PartialEq for VmValue {
+                panic!("Cannot perform comparison between arrays!")
+            }
+            (VmValue::f32Array4(val1), VmValue::f32Array4(val2)) => {
+                panic!("Cannot perform comparison between arrays!")
+            }
+            (VmValue::f32Array8(val1), VmValue::f32Array8(val2)) => {
+                panic!("Cannot perform comparison between arrays!")
+            }
+            (VmValue::f64Array2(val1), VmValue::f64Array2(val2)) => {
+                panic!("Cannot perform comparison between arrays!")
+            }
+            (VmValue::f64Array4(val1), VmValue::f64Array4(val2)) => {
+                panic!("Cannot perform comparison between arrays!")
+            }
+            (VmValue::f64Array8(val1), VmValue::f64Array8(val2)) => {
+                panic!("Cannot perform comparison between arrays!")
+            }
+            (_, _) => panic!("Cannot perform compare between different variants!"),
+        }
+    }
+}
+
 impl Add for VmValue {
-    type Output = VmValue;
+    type Output = Result<VmValue, VmError>;
 
     // TODO: Possibly use native SIMD for arrays, but benchmark first
-    fn add(self, other: VmValue) -> VmValue {
+    fn add(self, other: VmValue) -> Result<VmValue, VmError> {
         match (self, other) {
-            (VmValue::I32(val1), VmValue::I32(val2)) => VmValue::I32(val1 + val2),
-            (VmValue::I64(val1), VmValue::I64(val2)) => VmValue::I64(val1 + val2),
-            (VmValue::F32(val1), VmValue::F32(val2)) => VmValue::F32(val1 + val2),
-            (VmValue::F64(val1), VmValue::F64(val2)) => VmValue::F64(val1 + val2),
+            (VmValue::I32(val1), VmValue::I32(val2)) => match val1.checked_add(val2) {
+                Some(result) => Ok(VmValue::I32(result)),
+                None => Err(VmError::Overflow),
+            },
+            (VmValue::I64(val1), VmValue::I64(val2)) => match val1.checked_add(val2) {
+                Some(result) => Ok(VmValue::I64(result)),
+                None => Err(VmError::Overflow),
+            },
+            (VmValue::F32(val1), VmValue::F32(val2)) => match VmValue::sum_f32(&val1, &val2) {
+                Some(result) => Ok(VmValue::F32(result)),
+                None => Err(VmError::Infinity),
+            },
+            (VmValue::F64(val1), VmValue::F64(val2)) => match VmValue::sum_f64(&val1, &val2) {
+                Some(result) => Ok(VmValue::F64(result)),
+                None => Err(VmError::Infinity),
+            },
             (VmValue::i32Array2(val1), VmValue::i32Array2(val2)) => {
                 let mut result: [i32; 2] = [0; 2];
-                let src = val1.iter().zip(&val2).map(|(a, b)| a + b);
+
+                let src = val1
+                    .iter()
+                    .zip(&val2)
+                    .map(|(a, b)| a.checked_add(*b).ok_or(VmError::Overflow));
 
                 for (r, v) in result.iter_mut().zip(src) {
-                    *r = v;
+                    *r = v.unwrap();
```

Comment:
This will panic if an overflow happens. It should instead propagate the error upward.